File: apps/recommendation-service/app/services/candidate_retrieval_service.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from app.database.recommendation_state_repository import RecommendationStateRepository


@dataclass
class RetrievalBatch:
    source: str
    candidates: list[dict[str, Any]]
    has_next: bool
    score_version: str


class CandidateRetrievalService:
    def __init__(self, repository: RecommendationStateRepository):
        self.repository = repository

    def get_semantic_online_batch(
        self,
        viewer_id: str,
        offset: int,
        size: int,
    ) -> RetrievalBatch:
        safe_offset = max(0, int(offset))
        safe_size = max(1, int(size))
        request_limit = safe_offset + safe_size + 1

        rows = self.repository.search_semantic_candidates(
            viewer_id=viewer_id,
            limit=request_limit,
            overscan=max(request_limit * 3, safe_size * 3),
        )
        page_rows = rows[safe_offset : safe_offset + safe_size + 1]

        return RetrievalBatch(
            source="semantic_online",
            candidates=[
                {
                    "candidateId": str(row["candidateId"]),
                    "candidateProfileText": row.get("candidateProfileText"),
                    "retrievalScore": float(row.get("retrievalScore", 0.0)),
                    "source": "semantic_online",
                }
                for row in page_rows[:safe_size]
            ],
            has_next=len(page_rows) > safe_size,
            score_version="recommendation-query-pipeline-v1",
        )

    def filter_graph_projection(
        self,
        viewer_id: str,
        candidates: list[dict[str, Any]],
    ) -> list[dict[str, Any]]:
        if not candidates:
            return []

        excluded_ids = self.repository.get_graph_excluded_candidate_ids(
            viewer_id,
            [str(candidate["candidateId"]) for candidate in candidates],
        )

        return [
            candidate
            for candidate in candidates
            if str(candidate["candidateId"]) not in excluded_ids
        ]

File: apps/search-recommendation-service/src/modules/recommendation/services/candidate-retrieval.service.ts
import { Injectable } from '@nestjs/common';
import { RecommendationStateRepository } from './recommendation-state.repository';

export interface RetrievalBatch {
  source: string;
  candidates: any[];
  hasNext: boolean;
  scoreVersion: string;
}

@Injectable()
export class CandidateRetrievalService {
  constructor(private readonly repository: RecommendationStateRepository) {}

  async getSemanticOnlineBatch(
    viewerId: string,
    offset: number,
    size: number,
  ): Promise<RetrievalBatch> {
    const safeOffset = Math.max(0, offset);
    const safeSize = Math.max(1, size);
    const requestLimit = safeOffset + safeSize + 1;

    const rows = await this.repository.searchSemanticCandidates(
      viewerId,
      [], // We pass empty vector as placeholder; actual embedding retrieval is handled inside repository
      requestLimit,
      Math.max(requestLimit * 3, safeSize * 3),
    );
    const pageRows = rows.slice(safeOffset, safeOffset + safeSize + 1);

    return {
      source: 'semantic_online',
      candidates: pageRows.slice(0, safeSize).map((row) => ({
        candidateId: String(row.candidateId),
        candidateProfileText: row.candidateProfileText,
        retrievalScore: Number(row.retrievalScore || 0.0),
        embedding: row.embedding,
        source: 'semantic_online',
      })),
      hasNext: pageRows.length > safeSize,
      scoreVersion: 'recommendation-query-pipeline-v1',
    };
  }

  async filterGraphProjection(
    viewerId: string,
    candidates: any[],
  ): Promise<any[]> {
    if (candidates.length === 0) return [];
    const excludedIds = await this.repository.getGraphExcludedCandidateIds(
      viewerId,
      candidates.map((c) => String(c.candidateId)),
    );

    return candidates.filter((c) => !excludedIds.has(String(c.candidateId)));
  }
}
